pydantic/_internal/_dataclasses.py

```diff
@@ -92,10 +95,25 @@ def prepare_dataclass(
 
     core_config = generate_config(config, cls)
     cls.__pydantic_fields__ = fields
-    cls.__pydantic_validator__ = SchemaValidator(schema, core_config)
+    cls.__pydantic_validator__ = validator = SchemaValidator(schema, core_config)
     # this works because cls has been transformed into a dataclass by the time "cls" is called
     cls.__pydantic_serializer__ = SchemaSerializer(schema, core_config)
 
+    extra = config.get('extra', Extra.ignore)  # ignore is the default in V1
+    if extra is Extra.allow:
+        raise ValueError(
+            'extra=Extra.allow is not allowed for dataclasses.'
+            ' Only Extra.ignore (the default) and Extra.forbid are allowed.'
+        )
+
+    if config.get('validate_assignment', False) is True:
```

Comment:
```suggestion
    if config.get('validate_assignment'):
```
I'm also fine keeping as is if you think that's clearer

Reply:
Good catch. I've been trying to use `get` because (1) it plays nicer with my type checker / IDE and (2) we've talked about getting rid of the magic defaults so I'm just not relying on them.